File: repos/sample-cs/src/Auth/Validator.cs
namespace SampleApp.Auth;

/// <summary>Kiểm tra format username + password.</summary>
public class Validator
{
    public virtual bool Validate(string username, string password)
    {
        return CheckUsername(username) && CheckPassword(password);
    }

    protected virtual bool CheckUsername(string username)
    {
        return !string.IsNullOrEmpty(username) && username.Length >= 4;
    }

    protected virtual bool CheckPassword(string password)
    {
        return !string.IsNullOrEmpty(password) && password.Length >= 8;
    }
}

/// <summary>Validator nghiêm ngặt cho admin.</summary>
public class StrictValidator : Validator
{
    protected override bool CheckPassword(string password)
    {
        if (string.IsNullOrEmpty(password) || password.Length < 12) return false;
        bool hasUpper = false, hasDigit = false, hasSymbol = false;
        foreach (var c in password)
        {
            if (char.IsUpper(c)) hasUpper = true;
            else if (char.IsDigit(c)) hasDigit = true;
            else if (!char.IsLetterOrDigit(c)) hasSymbol = true;
        }
        return hasUpper && hasDigit && hasSymbol;
    }
}

File: repos/sample-ts/src/auth/Validator.ts
/** Base validator — kiểm tra format thô của username/password. */
export class Validator {
  validate(username: string, password: string): boolean {
    return this.checkUsername(username) && this.checkPassword(password);
  }

  protected checkUsername(username: string): boolean {
    return username.length >= 4 && /^[a-zA-Z0-9_]+$/.test(username);
  }

  protected checkPassword(password: string): boolean {
    return password.length >= 8;
  }
}

/** Validator nghiêm ngặt hơn cho admin — yêu cầu password phức tạp. */
export class StrictValidator extends Validator {
  protected checkPassword(password: string): boolean {
    return (
      password.length >= 12 &&
      /[A-Z]/.test(password) &&
      /[0-9]/.test(password) &&
      /[^A-Za-z0-9]/.test(password)
    );
  }
}
